### ISSUE TITLE
support oct_uri as a remote data source for wasm

### ISSUE THREAD

**Issue by arkodg** [envoyproxy/envoy#33212]
*Title*: *One line description*

*Description*:

Envoy only supports fetching a WASM module [remotely via http](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#config-core-v3-remotedatasource)

Majority of the WASM Modules now live in OCI Registries, so would be great if an `oci` field was supported to allow users to fetch WASM modules directly instead of using a shim (an auxiliary control plane creating a temporary file server or a sidecar agent that prefetches from an OCI registry and persists it into the common filesystem) that deals with the CRUD state machine

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.
* OCIGetter in Helm (in Golang): https://github.com/helm/helm/blob/main/pkg/getter/ocigetter.go
* Istio's API for WASM https://istio.io/latest/docs/reference/config/proxy_extensions/wasm-plugin/

**Comment by yanavlasov** [comment]
Adding @lizan and @mpwarres as area owners

**Comment by kyessenov** [comment]
1. There are two formats, e.g. https://cloud.google.com/service-extensions/docs/prepare-plugin-code#container-image, https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/how-to-build-oci-images.md#overview.
2. How are secrets going to be distributed? Do you expect CDS+SDS for the remote?
3. Please fix HTTP and HTTPS URL fetch as well. https://github.com/envoyproxy/envoy/issues/29824

**Comment by jewertow** [comment]
> requested @zhaohuabing to look into https://github.com/envoyproxy/envoy/issues/29824 when he has some free cycles

@zhaohuabing did you already start or are you going to start working on this issue soon? If not, I could take care of it.

**Comment by zhaohuabing** [comment]
@jewertow I haven't started on this. Please feel free to move on if you have time.

**Comment by kyessenov** [comment]
If you're going to implement it, let's settle on the two design issues:
* Caching: we probably don't want to fetch 100MB by URL every single time a listener is loaded. That requires some caching strategy, with expiration.
* Reliability: how do we stage xDS and not apply a Wasm config until the blob is fetched? What happens if the blob fails to be fetched?

**Comment by jewertow** [comment]
@kyessenov are your suggestions relevant for HTTPS URI or for OCI? I was going to start with HTTPS URI and I didn't investigate the scope of OCI yet.

**Comment by kyessenov** [comment]
Yes, they are. OCI is not that different from a regular HTTPS. It's just a tar with structured content and some authentication  schemes.

**Comment by zhaohuabing** [comment]
Should HTTPS be similar to the current HTTP fetching and caching mechanism？

**Comment by kyessenov** [comment]
@zhaohuabing The current HTTP fetching in Wasm is somewhat broken, we warn people not to use it. If you use a cluster specifier, then HTTPS vs HTTP becomes a configuration in the upstream cluster, so yes, it should be similar.

**Comment by zhaohuabing** [comment]
> @zhaohuabing The current HTTP fetching in Wasm is somewhat broken, we warn people not to use it. If you use a cluster specifier, then HTTPS vs HTTP becomes a configuration in the upstream cluster, so yes, it should be similar.

@kyessenov Could you elaborate on what's broken in HTTP fetching? We plan to use HTTP Wasm code source in EG and want to avoid any known issues. Thanks!

**Comment by kyessenov** [comment]
> @kyessenov Could you elaborate on what's broken in HTTP fetching? We plan to use HTTP Wasm code source in EG and want to avoid any known issues. Thanks!

It crashes :) I don't have enough time to troubleshoot and fix it, but it needs to be fixed before using it in production.

**Comment by zhaohuabing** [comment]
After testing, I confirmed that HTTPS schema works for the Wasm HTTP code source. It seems that TLS is solely associated with the Cluster configuration and does not require handling within the Wasm filter.

https://github.com/zhaohuabing/playground/blob/2e2d0b49c36eda2a5282cbdeba482455b1b1109a/envoy/wasm/envoy.yaml#L28-L49

**Comment by jewertow** [comment]
@kyessenov I wanted to start working on support for OCI scheme, and since this requires to send a few consecutive requests (GET manifests -> GET blobs -> GET layers), I wonder if I should follow asynchronous pattern used in `RemoteDataFetcher`? I mean, should I implement fetchers extending `Http::AsyncClient::Callbacks` for manifests, blobs and layers and call them consecutively in `onSuccess`? WDYT?

**Comment by jewertow** [comment]
I implemented a simple prototype of OCI client and I was able to successfully run Envoy with WASM plugin fetched from the docker registry, but now I have more questions, because I found some issues.

1. What image format should be supported?
I started with [wasm-image-spec](https://github.com/solo-io/wasm-image-spec), but I didn't find a tool to push an image in this format, so I tried [wasm-to-oci](https://github.com/engineerd/wasm-to-oci) as its format works with many popular registries (including Docker Hub), but I am not sure what's the standard and which format Envoy should support.
2. Which registries should we support?
I read this [spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md) to understand how we should pull images, but I figured out that Docker requires authorization header even though an image is public, and that requires additional request for the oauth token, but that authentication request is not part of the distribution spec. I tested also quay.io and gcr.io, and these registries do not require additional authorization, so it seems that docker is not fully compliant.
3. How pull secret for private registries should be provided in the configuration? [GenericSecret](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/secret.proto#extensions-transport-sockets-tls-v3-genericsecret)?

@kyessenov @arkodg @zhaohuabing WDYT?
